Doc fix: remove render prop from new Form component examples (#7505)
marmelab/react-admin: docs/Upgrade.md

To upgrade, replace all occurrences of `<FormWithRedirect>` with  `<Form>`.

Also, the `render` prop has been removed, and the form content is now passed as child of the `<Form>` element.

```diff
- import { FormWithRedirect } from 'react-admin';
+ import { Form } from 'react-admin';

export const MyForm = () => (
-    <FormWithRedirect
-        render={() => ...}
-    />
+    <Form>

> +        ...

Reviewer: this isn't explicit enough. You should replace the `...` in the diff by a dummy component to make the change clearer

+    </Form>
);
```

If you need to access the form state (`valid`, `invalid`, `pristine`, `dirty`), you can call [the react-hook-form `useFormState` hook](https://react-hook-form.com/api/useformstate) instead:

```diff
import { FormWithRedirect } from 'react-admin';
+ import { Form } from 'react-admin';
+ import { useFormState } from 'react-hook-form';

const MyCustomForm = () => {
    return (
-        <FormWithRedirect
+        <Form>
-            render={({ valid, dirty, handleSubmit }) => (
-               <form onSubmit={handleSubmit}>
                    ...
-                    <SubmitButton disabled={!dirty || !valid}>Save</SubmitButton>
+                    <SubmitButton>Save</SubmitButton>
-               </form>
            )}
-       />
+       </Form>
    );
};

-const SubmitButton = ({ disabled, ...props }) => {
+const SubmitButton = (props) => {
+    const { isDirty, isValid } = useFormState();
    return (
-        <button disabled={disabled} {...props} />

> +        <button disabled={!isDirty || !isValid} {...props} />

Reviewer: this doesn't work - you need at least a type="submit"

Author: Isn't `type="submit"` the default for a button? I'll test it...

Author: I confirm it's working without `type="submit"`.

Do you want me to add it anyway? In this case, should I also add it to the "before" line, to make it clear that the diff is not about this?

```diff
const MyForm = () => {
    return (
-        <Create>
+        <Create redirect="show">
-            <FormWithRedirect
+            <Form>
-                render={({ handleSubmitWithRedirect, ...formProps }) => (
                         <TextInput source="name" />
-                        <MySaveButton handleSubmitWithRedirect={handleSubmitWithRedirect}>
+                        <MySaveButton />
-                )}
-            />
+            </Form>
        </Create>
    );
};

-const MySaveButton = ({ handleSubmitWithRedirect }) => (
+const MySaveButton = () => {
    return (
-        <button onClick={() => handleSubmitWithRedirect('show')}>Save</button>

> +        <button type="submit">Save</button>

Reviewer: I don't understand this change. This SaveButton has the defautl behavior - it's not customized

Author: This example was about showing how to use `handleSubmitWithRedirect` to redirect to the **show** page after submission.

So in the new example I encouraged to make the following change:
```diff
-        <Create>
+        <Create redirect="show">
```
And then you don't need to customize the Save button anymore.

What is troubling you? Did you miss the `redirect="show"` I added, or is it because the `redirect` prop is more limited than what you could do with `handleSubmitWithRedirect` and you would rather have me add a more complete example?

    );
);
```

> You can also use a function to handle the redirection target, as described in [this section](./EditTutorial.md#redirection-after-submission).

Reviewer: ```suggestion
You can also use a function to handle the redirection target, as described in [the redirection after submittion documentation](./EditTutorial.md#redirection-after-submission).
```

Don't use 'this thing' as support of a hyperlink, but rather the title of the this you link to.
